File: Assets/Scripts/testScene/touchTest.js
#pragma strict
var obj_eye : GameObject;
var startTouchPosition:Vector2;
var endTouchPosition:Vector2;
var canSetSpeed:boolean;
var isInBallRange:boolean;
var timeStart:float;
var timeEnd:float;
var mouseVelocity:Vector2;
var cam:Camera;
private var MAX_SPEED : float;


function Start () {
	canSetSpeed = false;
	isInBallRange = false;
	timeStart = 0.0;
	timeEnd = 0.0;
	MAX_SPEED = 20.0;
}

function Update () {
	obj_eye = GameObject.Find("obj_eye_ball");
	var touchCount = Input.touchCount;
	for(var i = 0; i < touchCount;i++)
	{	
		if(Input.GetTouch(i).phase == TouchPhase.Began)
		{
			startTouchPosition = Input.GetTouch(i).position;
			var hit:RaycastHit;
            // Construct a ray from the current touch coordinates
    		 var ray : Ray = cam.ScreenPointToRay (startTouchPosition);
   			 if (Physics.Raycast (ray, hit) && hit.transform.name.Substring(0,12) == obj_eye.name)isInBallRange = true;
			else continue;
			timeStart = Time.time;
		}
		if(Input.GetTouch(i).phase == TouchPhase.Ended)
		{
			endTouchPosition = Input.GetTouch(i).position;
			timeEnd = Time.time;
			if(isInBallRange){canSetSpeed = true;break;}
			
		}
	}
	var deltaTime = timeEnd - timeStart;
	if(canSetSpeed)
	{
		
		if(deltaTime > 0)
		{
			var deltaDistanceX = endTouchPosition.x - startTouchPosition.x;
			var deltaDistanceY = endTouchPosition.y - startTouchPosition.y;
			mouseVelocity.x = deltaDistanceX / (deltaTime * 100);
			mouseVelocity.y = deltaDistanceY / (deltaTime * 100);
		}
		var totalVelocity = mouseVelocity.x * mouseVelocity.x + mouseVelocity.y * mouseVelocity.y;
		var sqrTotalVelocity = Mathf.Sqrt(totalVelocity);
		if(totalVelocity > (MAX_SPEED * MAX_SPEED)){
			obj_eye.rigidbody.velocity = Vector3(mouseVelocity.x * MAX_SPEED / sqrTotalVelocity,0,mouseVelocity.y * MAX_SPEED / sqrTotalVelocity);
			//Debug.Log(gameObject.rigidbody.velocity);
			canSetSpeed = false;
		}
		else
		{
			obj_eye.rigidbody.velocity = Vector3(mouseVelocity.x,0,mouseVelocity.y);
			//Debug.Log(gameObject.rigidbody.velocity);
			canSetSpeed = false;
		}
	}
}

File: Assets/Scripts/testScene/touchTest.cs

using UnityEngine;
using System.Collections;

public class touchTest : MonoBehaviour {
	
	public Camera cam;
	private GameObject objBall;
	private Touch beginTouch;
	private TouchPhase currentPhase = TouchPhase.Canceled;
	private float maxForce;
	private float minForce;
	private Vector3 originalPosition;
	private int shootGestureTime;
	private string[] ballArray = new string[4]{"obj_eye_ball","obj_onion","obj_carrot","obj_steam_bread"};
	private const int maxShootGestureTime = 5;
	
	// Use this for initialization
	void Start () {
		maxForce = 400;
		minForce = 200;
		shootGestureTime = 0;
		originalPosition = this.transform.position;
		originalPosition.x = (originalPosition.x + 3.84f) * 100;
		originalPosition.z = (originalPosition.z + 5.12f) * 100;
		StartCoroutine(calcShootGestureTime());
	}
	
	void handleTouch(Touch touch) {
		//Debug.Log(touch.position);
			switch(currentPhase) {
			case TouchPhase.Canceled:
				if(touch.phase == TouchPhase.Began) {
					//Debug.Log(this.transform.position);
					RaycastHit hit = new RaycastHit();
            		// Construct a ray from the current touch coordinates
    		 		Ray ray = new Ray();
					ray = cam.ScreenPointToRay (touch.position);
   			 		if (!(Physics.Raycast (ray, out hit) && IsCountInArray(hit.transform.name)))
					{
						//Debug.Log(hit.transform.name);
						return;
					}
					//Debug.Log("touch begin");
					shootGestureTime = 0;
					beginTouch = touch;
					currentPhase = TouchPhase.Began;
				}
				break;
			case TouchPhase.Began: {
				if(touch.phase == TouchPhase.Ended && touch.fingerId == beginTouch.fingerId) {
					//Debug.Log("touch end");
					float deltaX = touch.position.x - beginTouch.position.x;
					float deltaY = touch.position.y - beginTouch.position.y;
					float distance = Mathf.Sqrt(deltaX * deltaX + deltaY * deltaY);
					if(distance > 0) {	
						//rotate
						float rotationY = Mathf.Acos((beginTouch.position.y - touch.position.y) / distance) / Mathf.PI * 180;
						if(touch.position.x <= beginTouch.position.x)
							rotationY = 180 + rotationY;
						else
							rotationY = 180 - rotationY;
						
						//Debug.Log("deltaX:"+deltaX+"deltaY:"+deltaY+"Y:"+rotationY);
						objBall.transform.rotation = Quaternion.AngleAxis(rotationY, Vector3.up);
						
						//Debug.Log("shoot gesture time:"+shootGestureTime);
						float ratio = 0;
						if(shootGestureTime < maxShootGestureTime) {
							ratio = (float)(maxShootGestureTime - shootGestureTime) / (float)maxShootGestureTime;
						}
						float forceCoefficient = (maxForce - minForce) * ratio + minForce;
						//forceCoefficient = 400;
						//Debug.Log("forceCoefficient:"+forceCoefficient);
						Vector3 force = new Vector3(deltaX / distance * forceCoefficient, 1, deltaY / distance * forceCoefficient);
						objBall.rigidbody.AddForce(force);	
						
						currentPhase = TouchPhase.Canceled;	
					}
					else {
						currentPhase = TouchPhase.Canceled;
					}
				}
				break;
			}
			default:
				break;
			}
	}

	private IEnumerator calcShootGestureTime() {
		while(true) {
			shootGestureTime += 1;
			yield return new WaitForSeconds(0.1f);
		}
	}
	
	// Update is called once per frame
	void FixedUpdate () {
		if (Input.touchCount != 0) {
			Touch touch;
			for(int i = 0; i < Input.touchCount; i++) {
				touch = Input.touches[i];
				this.handleTouch(touch);
			}
		}
	}
	
	bool IsCountInArray(string obj)
	{
		for(int i = 0;i < 4;i++)
		{
			if(obj == ballArray[i]){
			objBall = GameObject.Find(ballArray[i]);
			return true;
			}
		}
		return false;
	}
}
